flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java

```diff
@@ -1081,4 +1269,345 @@ public void setAttribute(Integer vertexId, Attribute attribute) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public String toString() {
+        return "StreamGraph(jobId: " + jobId + ")";
+    }
+
+    /**
+     * A static inner class designed to hold user-defined objects for serialization and
+     * deserialization in the stream graph.
+     */
+    private class UserDefinedObjectsHolder implements Serializable {
+
+        private static final long serialVersionUID = 1L;
+
+        private final SerializedValue<
+                        Map<Integer, Tuple3<Integer, StreamPartitioner<?>, StreamExchangeMode>>>
+                serializedVirtualPartitionNodes;
+
+        private final SerializedValue<ExecutionConfig> serializedExecutionConfig;
+
+        private SerializedValue<Map<Integer, StreamNode>> serializedStreamNodes;
+
+        /**
+         * This collection stores operator factories serialized separately from the {@link
+         * StreamGraph}. This separation allows for the parallel serialization of operator
+         * factories, improving the overall performance of the serialization process.
+         *
+         * <p>Each tuple in this collection consists of an integer key that identifies the stream
+         * node, and a value that wraps the serialized representation of the associated {@link
+         * StreamOperatorFactory} instance.
+         */
+        private Collection<Tuple2<Integer, SerializedValue<StreamOperatorFactory<?>>>>
+                streamNodeToSerializedOperatorFactories;
+
+        private final SerializedValue<Map<Integer, Tuple2<Integer, OutputTag>>>
+                serializedVirtualSideOutputNodes;
+
+        public UserDefinedObjectsHolder(
+                Map<Integer, StreamNode> streamNodes,
+                Map<Integer, Tuple2<Integer, OutputTag>> virtualSideOutputNodes,
+                Map<Integer, Tuple3<Integer, StreamPartitioner<?>, StreamExchangeMode>>
+                        virtualPartitionNodes,
+                ExecutionConfig executionConfig,
+                @Nullable StateBackend stateBackend,
+                @Nullable CheckpointStorage checkpointStorage,
+                Executor serializationExecutor)
+                throws IOException {
+            serializeStreamNodes(streamNodes, serializationExecutor);
+
+            this.serializedVirtualSideOutputNodes = new SerializedValue<>(virtualSideOutputNodes);
+            this.serializedVirtualPartitionNodes = new SerializedValue<>(virtualPartitionNodes);
+            this.serializedExecutionConfig = new SerializedValue<>(executionConfig);
+
+            if (stateBackend != null && serializedStateBackend == null) {
+                serializedStateBackend = new SerializedValue<>(stateBackend);
+            }
+
+            if (checkpointStorage != null && serializedCheckpointStorage == null) {
+                serializedCheckpointStorage = new SerializedValue<>(checkpointStorage);
+            }
```

Comment:
How about to ensure `JobCheckpointingSettings` is created right after `stateBackend` and `checkpointStorage` are set, and let it be the holder of `serializedStateBackend` and `serializedCheckpointStorage`? Looks to me it can simplify things a lot.